client_test.go

```diff
@@ -200,3 +204,175 @@ func TestClientRunRawRequestNoURL(t *testing.T) {
 		t.Errorf("Unexpected error response: %s", err.Error())
 	}
 }
+
+func TestAuthTokenAccessors(t *testing.T) {
+	c := APIClient{}
+
+	if c.GetAuthToken() != nil {
+		t.Error("Empty client should return a nil auth token")
+	}
+
+	token := &redfish.AuthToken{}
+	c.SetAuthToken(token)
+
+	if c.GetAuthToken() != token {
+		t.Error("Client should return current auth token")
+	}
+
+	c.SetAuthToken(nil)
+
+	if c.GetAuthToken() != nil {
+		t.Error("Should return nil")
+	}
+}
+
+const brokenHuaweiServiceRoot = `{
+  "@odata.context": "/redfish/v1/$metadata#ServiceRoot",
+  "@odata.id": "/redfish/v1/",
+  "@odata.type": "#ServiceRoot.v1_1_0.ServiceRoot",
+  "Id": "RootService",
+  "Name": "Root Service",
+  "RedfishVersion": "1.0.2",
+  "UUID": "2D0B7460-48DC-BB02-EB11-592730502FEC",
+  "Systems": {
+    "@odata.id": "/redfish/v1/Systems"
+  },
+  "Chassis": {
+    "@odata.id": "/redfish/v1/Chassis"
+  },
+  "Managers": {
+    "@odata.id": "/redfish/v1/Managers"
+  },
+  "Tasks": {
+    "@odata.id": "/redfish/v1/TaskService"
+  },
+  "SessionService": {
+    "@odata.id": "/redfish/v1/SessionService"
+  },
+  "AccountService": {
+    "@odata.id": "/redfish/v1/AccountService"
+  },
+  "EventService": {
+    "@odata.id": "/redfish/v1/EventService"
+  },
+  "UpdateService": {
+    "@odata.id": "/redfish/v1/UpdateService"
+  },
+  "Registries": {
+    "@odata.id": "/redfish/v1/Registries"
+  },
+  "JsonSchemas": {
+    "@odata.id": "/redfish/v1/JSONSchemas"
+  },
+  "Oem": {
+    "Huawei": {
+      "SmsUpdateService": null,
+      "ProductName": "2288H V5",
+      "HostName": "2102311XBxxxxxxxxxxx",
+      "LanguageSet": "en,zh,ja,fr",
+      "Copyright": "Huawei Technologies Co., Ltd. 2004-2020. All rights reserved.",
+      "DomainName": null,
+      "AccountLockoutDuration": 300
+    }
+  }
+}`
+
+func huaweiPatch(s *Service) error {
+	// check if this is a Huawei BMC
+	if s.Vendor != "Huawei" && s.Vendor != "" {
+		return nil
+	}
+	var result map[string]any
+	err := json.Unmarshal(s.Oem, &result)
+	if err != nil {
+		return err
+	}
+	var isHuawei = false
+	_, isHuawei = result["Huawei"]
+	if !isHuawei {
+		return nil
+	}
+
+	// This code would be called from an external package
+	// => can only use public API!
+	if s.sessions == "" {
+		s.sessions = "/redfish/v1/SessionService/Sessions"
+	}
```

Comment:
Not sure about how to solve this one.

This feels too specific if we added now a Setter for `Service.sessions`.

Another approach would be to have a hook just before the decoding of the JSON response.

What do you think?